### PR TITLE
feat: enhancement the app can enable random status code

### PR DESCRIPTION
**Pull request by Thanakorn-ki**
Relate issue: #269 

# Feature
Enhance the random status code feature. When checked enable random status code the app will random status from the create response. 

image

# Screenshot

Guideline enable status

### CODE REVIEW

**Review by 255kb — CHANGES_REQUESTED** [review]
Hi @Thanakorn-ki Thank you for contributing. 
Could you please comment on the #269 issue so I can assign it to you? 

Also, I would like to discuss the placement of the checkbox. I think this should not be at the top with the route options but next to the responses list. Maybe we could remove the text from the "Add response" button and put it in a tooltip, and add another button with an icon, which could be toggled on/off. 
But I need to find an icon as the shuffle icon (the two crossed arrows) from the material design icons pack has already been used for the CORS options. 
image

I also have several remarks concerning the implementation:
- I would rename `enableRandomStatusCode` to `randomResponse` as we are not really randomizing a "status code" inside a response but rather a response inside a route. I hope it makes sense. 
- Also we will need a new migration no. 13 in the migration lib + a migration test in `test/data/migrations/13`
- I saw you update all the environment test files, and I thank you for this, I know it's a pain and not really straightforward. I would like at one point to create a small script to take care of this, but I didn't have time yet :) However, after you create the migration no. 13 you will also have to update the `lastMigration` property in all these files. And last thing, you shouldn't add the new property `enableRandomStatusCode` (or `randomReponse`) to the test files that are in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`. These test environments files are here to test migrations/imports from old versions so they shouldn't be modified at all. 

Thank you for your time and help :)

**Comment by Thanakorn-ki** [comment]
> Hi @Thanakorn-ki Thank you for contributing.
> Could you please comment on the #269 issue so I can assign it to you?
> 
> Also, I would like to discuss the placement of the checkbox. I think this should not be at the top with the route options but next to the responses list. Maybe we could remove the text from the "Add response" button and put it in a tooltip, and add another button with an icon, which could be toggled on/off.
> But I need to find an icon as the shuffle icon (the two crossed arrows) from the material design icons pack has already been used for the CORS options.
> image
> 
> I also have several remarks concerning the implementation:
> 
> * I would rename `enableRandomStatusCode` to `randomResponse` as we are not really randomizing a "status code" inside a response but rather a response inside a route. I hope it makes sense.
> * Also we will need a new migration no. 13 in the migration lib + a migration test in `test/data/migrations/13`
> * I saw you update all the environment test files, and I thank you for this, I know it's a pain and not really straightforward. I would like at one point to create a small script to take care of this, but I didn't have time yet :) However, after you create the migration no. 13 you will also have to update the `lastMigration` property in all these files. And last thing, you shouldn't add the new property `enableRandomStatusCode` (or `randomReponse`) to the test files that are in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`. These test environments files are here to test migrations/imports from old versions so they shouldn't be modified at all.
> 
> Thank you for your time and help :)

Thank you for your suggestion and I will fix it.

**Comment by Thanakorn-ki** [comment]
@255kb 
Are you ok this?
When we want to shuffle response this icon adds the color green and then we don't shuffle the default color of the app.

image

**Comment by Thanakorn-ki** [comment]
> @255kb
> Are you ok this?
> When we want to shuffle response this icon adds the color green and then we don't shuffle the default color of the app.
> 
> image

Or check icon ?

image

**Comment by 255kb** [comment]
@Thanakorn-ki I like the green/not green icon, but I wonder would prefer to use the blue color when activated, so it's consistent with the rest of the app. Also, could you please keep the two buttons and the dropdown part of the `btn-group` so there is no gap between them?
Concerning this icon, it's the perfect one for symbolizing randomness, but it is used for the CORS option right now. Maybe we should find something else for the CORS from the icons set https://material.io/resources/icons/?style=baseline
But I wonder what?

**Comment by 255kb** [comment]
I also think we should say more in the tooltip, maybe "Random response (will disable the rules)", and add an orange message inside the rules saying "Activating random responses disable the rules"
The message could be similar to the one above the body when a file is selected:
image

**Comment by Thanakorn-ki** [comment]
> I also think we should say more in the tooltip, maybe "Random response (will disable the rules)", and add an orange message inside the rules saying "Activating random responses disable the rules"
> The message could be similar to the one above the body when a file is selected:
> image

OK, it makes sense

**Comment by 255kb** [comment]
Don't bother changing the CORS icon for now, use the "shuffle" icon for your feature and I will do the changes to the CORS option after.

**Comment by Thanakorn-ki** [comment]
@255kb 
Last implement.

image

File: src/app/app.component.html [hunk 1]
```diff
@@ -286,8 +287,15 @@
                           <!-- Responses dropdown -->
                           <div *ngIf="activeRouteResponse$ | async as activeRouteResponse" class="btn-group dropdown mr-1" ngbDropdown>
                             <button type="button" class="btn btn-custom" (click)="addRouteResponse()">
-                              <i class="material-icons">add</i> <span class="d-none d-xl-inline">Add response</span>
+                              <i class="material-icons" ngbTooltip="Add response">add</i>
+                            </button>
+
+                            <button type="button" class="btn btn-custom" (click)="toggleRandomResponse(activeRoute.uuid)">
+                              <i
+                                [ngClass]="{'text-primary': getRandomResponse()}"
```
Comment (255kb) [inline src/app/app.component.html:295]: you should directly use `activeRoute.randomResponse` which is available there through an `async` pipe line 238:

```suggestion
                                [ngClass]="{'text-primary': activeRoute.randomResponse}"
```

File: src/app/app.component.html [hunk 2]
```diff
@@ -448,6 +456,11 @@
                                 <span *ngIf="activeRouteResponseLastLog$ | async as activeRouteResponseLastLog" class="btn btn-link btn-icon text-primary view-body-link" (click)="goToRouteResponseLog(activeRouteResponseLastLog.UUID)">View last body sent</span>
                               </ng-container>
                             </ng-container>
+
+                            <ng-container subheading>
+                              <ng-container *ngIf="getRandomResponse()"><span class="text-warning">Activating random responses disable the rules</span></ng-container>
+                            </ng-container>
+
```
Comment (255kb) [inline src/app/app.component.html:463]: I'm sorry, maybe what I suggested wasn't clear at all, but this message shouldn't be here which is reserved for body related messages. Instead, put it inside the "Rules" tab content at the top. I only suggested that you use the same design  than the body message. 
Something like 
```
<div class="text-warning text-center">Activating random responses disable the rules</div>
```

File: src/app/app.component.ts [hunk 3]
```diff
@@ -664,4 +664,20 @@ export class AppComponent implements OnInit, AfterViewInit {
...
+      toggleRandomResponse(uuid)
+    );
+  }
+
+  /**
+   * get Enable/disable random response
+   */
+  public getRandomResponse(): boolean {
+    return this.store.getActiveRoute().randomResponse;
+  }
```
Comment (255kb) [inline src/app/app.component.ts:682]: Remove as you will directly use `activeRoute` in the template.

(As a general rule, method calls should be avoided inside templates as they can have bad performances. This is mainly due to the change detection which will call the method each time something change even if the data returned by the method didn't change)

File: src/app/stores/actions.ts [hunk 4]
```diff
@@ -438,6 +439,16 @@ export function updateUIStateAction(properties: UIStateProperties) {
...
 
+/**
+ * Enable/disable random response
+ */
+export function toggleRandomResponse(uuid: string) {
+  return <const>{
+    type: ActionTypes.TOGGLE_RANDOM_RESPONSE,
+    uuid
+  };
+}
```
Comment (255kb) [inline src/app/stores/actions.ts:450]: You don't need to create a new action as you can use `updateRouteAction()`

File: src/app/stores/actions.ts [hunk 5]
```diff
@@ -51,7 +51,8 @@ export const enum ActionTypes {
   REMOVE_TOAST,
   SET_USER_ID,
   UPDATE_SETTINGS,
-  UPDATE_UI_STATE
+  UPDATE_UI_STATE,
+  TOGGLE_RANDOM_RESPONSE
```
Comment (255kb) [inline src/app/stores/actions.ts:55]: You don't need to create a new action as you can use updateRouteAction()

File: src/app/app.component.ts [hunk 6]
```diff
@@ -664,4 +664,20 @@ export class AppComponent implements OnInit, AfterViewInit {
   public duplicateRouteResponse() {
     this.environmentsService.duplicateRouteResponse();
   }
+
+  /**
+   * Enable/disable random response
+   */
+  public toggleRandomResponse(uuid: string) {
+    this.store.update(
+      toggleRandomResponse(uuid)
```
Comment (255kb) [inline src/app/app.component.ts:673]: please use `updateRouteAction()`
```suggestion
  public toggleRandomResponse(randomResponse: boolean) {
    this.store.update(
      updateRouteAction({ randomResponse: !randomResponse })
```

File: src/app/stores/reducer.ts [hunk 7]
```diff
@@ -907,6 +907,38 @@ export function environmentReducer(
...
+        return environment;
+      });
+
+      newState = {
+        ...state,
+        environments: newEnvironments,
+      };
+
+      break;
+    }
```
Comment (255kb) [inline src/app/stores/reducer.ts:940]: Not needed if you use `updateRouteAction`

File: test/data/import/new/full-export.json [hunk 8]
```diff
@@ -17,6 +17,7 @@
                         "endpoint": "answer",
                         "documentation": "",
                         "enabled": true,
+                        "randomResponse": false,
```
Comment (255kb) [inline test/data/import/new/full-export.json:20]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/import/new/route-export.json [hunk 9]
```diff
@@ -9,6 +9,7 @@
               "endpoint": "answer",
               "documentation": "",
               "enabled": true,
+              "randomResponse": false,
```
Comment (255kb) [inline test/data/import/new/route-export.json:12]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/import/old/1.5.1.json [hunk 10]
```diff
@@ -16,6 +16,7 @@
                     "documentation": "",
                     "method": "get",
                     "endpoint": "answer",
+                    "randomResponse": false,
```
Comment (255kb) [inline test/data/import/old/1.5.1.json:19]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/import/old/1.6.0-single-env.json [hunk 11]
```diff
@@ -15,6 +15,7 @@
               "documentation": "",
               "method": "get",
               "endpoint": "",
+              "randomResponse": false,
```
Comment (255kb) [inline test/data/import/old/1.6.0-single-env.json:18]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/import/old/1.6.0-single-route.json [hunk 12]
```diff
@@ -8,6 +8,7 @@
       "documentation": "",
       "method": "get",
       "endpoint": "",
+      "randomResponse": false,
```
Comment (255kb) [inline test/data/import/old/1.6.0-single-route.json:11]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/import/old/1.6.0.json [hunk 13]
```diff
@@ -16,22 +16,23 @@
                     "documentation": "",
                     "method": "get",
                     "endpoint": "answer",
+                    "randomResponse": false,
```
Comment (255kb) [inline test/data/import/old/1.6.0.json:19]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/migrations/10/environments.json [hunk 14]
```diff
@@ -13,6 +13,7 @@
         "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
         "documentation": "Test documentation",
         "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/10/environments.json:16]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/migrations/11/environments.json [hunk 15]
```diff
@@ -13,6 +13,7 @@
         "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
         "documentation": "Test documentation",
         "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/11/environments.json:16]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/migrations/12/environments.json [hunk 16]
```diff
@@ -13,6 +13,7 @@
         "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
         "documentation": "Test documentation",
         "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/12/environments.json:16]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/migrations/13/environments.json [hunk 17]
```diff
@@ -0,0 +1,51 @@
+[
+  {
+    "uuid": "323a25c6-b196-4d27-baf8-8aeb83d87c76",
+    "lastMigration": 13,
```
Comment (255kb) [inline test/data/migrations/13/environments.json:4]: This migration test should start from 12

File: test/data/migrations/13/environments.json [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    "latency": 0,
+    "port": 3000,
+    "routes": [
+      {
+        "uuid": "dd07b407-df1f-4393-83f2-7a832fdd6f99",
+        "method": "get",
+        "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
+        "documentation": "Test documentation",
+        "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/13/environments.json:16]: Do not add `randomResponse` property as we actually need to test the migration

File: test/data/migrations/9/environments.json [hunk 19]
```diff
@@ -13,6 +13,7 @@
         "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
         "documentation": "Test documentation",
         "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/9/environments.json:16]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/migrations/incompatible/environments.json [hunk 20]
```diff
@@ -13,6 +13,7 @@
         "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
         "documentation": "Test documentation",
         "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/incompatible/environments.json:16]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

File: test/data/migrations/pre-1.7.0/environments.json [hunk 21]
```diff
@@ -12,6 +12,7 @@
         "endpoint": "*/:var/a(b)?c/[0-9]{1,5}",
         "documentation": "Test documentation",
         "enabled": true,
+        "randomResponse": false,
```
Comment (255kb) [inline test/data/migrations/pre-1.7.0/environments.json:15]: Please do not modify files in `test/data/migrations`, `test/data/import/new` and `test/data/import/old`

**Review by 255kb — CHANGES_REQUESTED** [review]
Thank you for the changes, I made some comments. 
I'm sorry for the confusion on the yellow warning message, and for all the requested changes on the tests. I know it's a pain 😆 
If you want I can take it from here for the last details, let me know 👍

**Review by 255kb — APPROVED** [review]
Thank you for the modification.
I will take it from here and change the CORS icon. 

Thank you again for your contribution 🥳

**Pull request merged**